docs/rfcs/0003-sql-datastore.md

```diff
@@ -0,0 +1,281 @@
+-------------------------------------------------------------
+ Seq Scan on applications  (cost=0.00..1.04 rows=1 width=32)
+   Filter: ((data ->> 'name'::text) ~~ 'app-%'::text)
+(2 rows)
+```
+
+To resolve this issue, MySQL provides `FULLTEXT` index and PostgreSQL provides `GIN` index.
+
+For MySQL, the `FULLTEXT` index requires a physically stored column (`STORED GENERATED COLUMN` in case of JSON attribute indexes) to be used, which means __if we want to add it to a not yet existed column/JSON attribute, we have to use ALTER TABLE statement__. Otherwise, an error will be raised as follow
+```sql
+# MySQL v8
+mysql> CREATE FULLTEXT INDEX idz ON applications ((CAST(data->>'$.name' AS CHAR(10))));
+ERROR 3759 (HY000): Fulltext functional index is not supported.
+```
+Another option is to create a `stored generated column` that shadows value from an extra attribute of JSON data, then we could create indexes (FULLTEXT or normal) on that column for search features.
+
+For PostgreSQL, looks like we do not have an critical issue which this feature
+```sql
+postgres=> CREATE INDEX idf ON test.applications USING GIN(data);
+CREATE INDEX
+postgres=> \d test.applications
+               Table "test.applications"
+   Column   |  Type   | Collation | Nullable | Default 
+------------+---------+-----------+----------+---------
+ id         | uuid    |           | not null | 
+ project_id | uuid    |           | not null | 
+ data       | jsonb   |           | not null | 
+ disabled   | boolean |           | not null | 
+ created_at | bigint  |           | not null | 
+ updated_at | bigint  |           | not null | 
+Indexes:
+    "applications_pkey" PRIMARY KEY, btree (id)
+    "idf" gin (data)
+```
+
+MySQL: 👎 PostgreSQL: 👍
+
+## Performance of query operation on indexed JSON field
+
+For queries which uses search function on indexed JSON fields and without using JOIN (in our use-case)
+
+- For read queries, MySQL has a bit advantage due to its fast read-only characteristic. Besides, in case all virtual generated columns are secondary indexed columns, generated column values are materialized in the records of the index, which means MySQL will not recalculate virtual generated columns on query.
+- For write queries, PostgreSQL has a bit advantage due to MySQL cost on calculating virtual generated columns on each writes.
+
+MySQL: 👍 PostgreSQL: 👍
+## Able to keep advantage of schemaless pattern
+
+Yes, for both 🎉
+
+# Unresolved questions
+
+Currently, we have 2 points which need to investigate more
+## The support of each cloud providers
+
+With GCP and AWS (for now), since it's not the native service of those cloud providers, we have to use them fully-managed SQL services (SQL for GCP and RDS for AWS).\
+Besides, both AWS and GCP support the latest versions of PostgreSQL(v13) and MySQL(v8), but other cloud providers such as Azure only support up-to-date MySQL(v8) and a little behind PostgreSQL(v11).
+
+MySQL: 👍 PostgreSQL: 👎
```

Comment:
I think we should unify the way of using those emojis.
(👎  means bad, 👍  means good, 👍 👍  means very good.)

Maybe the latest PostgreSQL will be supported soon, so instead of 👎 , it should be 👍  for PostgreSQL and 👍 👍  for MySQL.

Reply:
I see 🙏 , updated by [`7c0b806` (#1612)](https://github.com/pipe-cd/pipe/pull/1612/commits/7c0b806b8971e8cb70862192d85ee479b7bd4d34)